### ISSUE TITLE
Error "no log driver named xxx" shown when using logging driver plugin in docker daemon

### ISSUE THREAD

**Issue by bin3377** [moby/moby#35553]
**Description**

The logging driver plugin can work with `docker run` but fail to startup when setting in docker daemon

**Steps to reproduce the issue:**
1. Install logging driver plugin
2. Verify it can work with docker run
3. Stop the docker daemon and start it again with same `--log-driver` and `--log-opt` parameters

**Describe the results you received:**
Docker daemon failed to start with message `no log driver named xxx is registered`
```
ec2-user@ip-10-0-0-37 ~]$ sudo docker plugin install store/sumologic/docker-logging-driver:1.0.2 --alias sumologic
Plugin "store/sumologic/docker-logging-driver:1.0.2" is requesting the following privileges:
 - network: [host]
Do you grant the above permissions? [y/N] y
1.0.2: Pulling from store/sumologic/docker-logging-driver
9c636ccd87f8: Download complete
Digest: sha256:54bc47ba8fa25ceeacf1cfa55f9f0a7c58e71923b57803fa684d7786e6910efb
Status: Downloaded newer image for store/sumologic/docker-logging-driver:1.0.2
Installed plugin store/sumologic/docker-logging-driver:1.0.2
[ec2-user@ip-10-0-0-37 ~]$ sudo docker plugin ls
ID                  NAME                DESCRIPTION                 ENABLED
5099889ff850        sumologic:latest    Sumo Logic logging driver   true
[ec2-user@ip-10-0-0-37 ~]$ docker run --log-driver=sumologic --log-opt sumo-url=https://stag-events.sumologic.net/receiver/v1/http/ZaVnC4dhaV03C98LzkuZtZpESfA5xLrvctBbTPkTZXVSQHtsLBkKNV3XrX7ePrqMyMijb41giONj_Bfdqxa8U-o_DpnZnxENs7K-1FYx13E5OEkkb3MN1Q== -d hello-world
8fe5243ef9bec81336e25ced3507bb0cf867998639503137fdffdc8741bb2c09
[ec2-user@ip-10-0-0-37 ~]$ sudo service docker stop
Stopping docker:                                           [  OK  ]
[ec2-user@ip-10-0-0-37 ~]$ sudo dockerd --log-driver=sumologic --log-opt sumo-url=https://stag-events.sumologic.net/receiver/v1/http/ZaVnC4dhaV03C98LzkuZtZpESfA5xLrvctBbTPkTZXVSQHtsLBkKNV3XrX7ePrqMyMijb41giONj_Bfdqxa8U-o_DpnZnxENs7K-1FYx13E5OEkkb3MN1Q==
Failed to set log opts: logger: no log driver named 'sumologic' is registered
```

**Describe the results you expected:**
The Docker daemon should accept the logging driver as in docker run

**Additional information you deem important (e.g. issue happens only occasionally):**
Always happen
Also tried configure with `/etc/sysconfig/docker` or `/etc/docker/daemon.json`, got same error when daemon startup

**Output of `docker version`:**

```
Client:
 Version:      17.06.2-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   3dfb8343b139d6342acfd9975d7f1068b5b1c3d3
 Built:        Tue Oct 17 18:33:39 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   402dd4a/17.06.2-ce
 Built:        Tue Oct 17 18:34:10 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 22
 Running: 0
 Paused: 0
 Stopped: 22
Images: 3
Server Version: 17.06.2-ce
Storage Driver: devicemapper
 Pool Name: docker-docker--pool
 Pool Blocksize: 524.3kB
 Base Device Size: 10.74GB
 Backing Filesystem: ext4
 Data file:
 Metadata file:
 Data Space Used: 444.1MB
 Data Space Total: 23.33GB
 Data Space Available: 22.88GB
 Metadata Space Used: 1.057MB
 Metadata Space Total: 25.17MB
 Metadata Space Available: 24.11MB
 Thin Pool Minimum Free Space: 2.333GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6e23458c129b551d5c9871e5174f6b1b7f6d1170
runc version: 810190ceaa507aa2727d7ae6f4790c76ec150bd2
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.58-18.51.amzn1.x86_64
Operating System: Amazon Linux AMI 2017.09
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 993.5MiB
Name: ip-10-0-0-37
ID: 3LPT:3A5U:RNOU:XN3V:2FWA:R5H6:SNF4:GGRD:73OF:LR4I:O6CI:75MV
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
AWS ECS AMI 2017.09.a

**Comment by fritzdj** [comment]
Poke @anusha-ragunathan. Is this something that you could help out with?

**Comment by aidansteele** [comment]
I was able to replicate the behaviour reported by @bin3377. I played around with it a bit further and found some strange results.

If you set `/etc/docker/daemon.json` to have `{"log-driver": "sumologic"}`, the Docker daemon starts successfully and `docker info` will report `sumologic` as the default logging driver. 

However, if you add `"log-opts": {"sumo-url": "http://example.com"}` to the daemon.json file it reports that `no log driver named sumologic is registered`. Likewise if I try set only the log-driver in daemon.json and log-opts on the command line.

**Comment by cpuguy83** [comment]
@aidansteele Thanks for the repro.
Sounds like this happens because the log-opt validator runs before plugins are registered.

**Comment by cpuguy83** [comment]
Fixed in #36306

**Comment by aidansteele** [comment]
Awesome turn-around time! Thanks for the great work
